src/utils/storage.ts

```diff
@@ -1,113 +1,386 @@
+ * @param storageTypes Full list of storage types allowed for this encoding.
+ * @param nestedSlotOffset Only used for structs. Since slots for struct members are 0-indexed, we
+ * need to be keeping track of the slot offset of the parent struct to figure out the final slot.
+ * See https://docs.soliditylang.org/en/v0.8.6/internals/layout_in_storage.html#mappings-and-dynamic-arrays for additional info.
+ * @param baseSlotKey Only used for maps. Keeps track of the base slot that other elements of the
+ * mapping need to work off of.
+ * @returns Variable encoded as a series of key/value slot pairs.
+ */
+function encodeVariable(
+  variable: any,
+  storageObj: SolidityStorageObj,
+  storageTypes: {
+    [name: string]: SolidityStorageType;
+  },
+  nestedSlotOffset = 0,
+  baseSlotKey?: string
+): StorageSlotPair[] {
+  let slotKey: string =
+    '0x' +
+    remove0x(
+      BigNumber.from(baseSlotKey || nestedSlotOffset)
+        .add(BigNumber.from(parseInt(storageObj.slot, 10)))
+        .toHexString()
+    ).padStart(64, '0');
+
+  const variableType = storageTypes[storageObj.type];
+  if (variableType.encoding === 'inplace') {
+    if (variableType.label === 'address' || variableType.label.startsWith('contract')) {
+      if (!ethers.utils.isAddress(variable)) {
+        throw new Error(`invalid address type: ${variable}`);
+      }
+
+      return [
+        {
+          key: slotKey,
+          val: padNumHexSlotValue(variable, storageObj.offset),
+        },
+      ];
+    } else if (variableType.label === 'bool') {
+      // Do some light parsing here to make sure "true" and "false" are recognized.
+      if (typeof variable === 'string') {
+        if (variable === 'false') {
+          variable = false;
+        }
+        if (variable === 'true') {
+          variable = true;
+        }
+      }
+
+      if (typeof variable !== 'boolean') {
+        throw new Error(`invalid bool type: ${variable}`);
+      }
+
+      return [
+        {
+          key: slotKey,
+          val: padNumHexSlotValue(variable ? '1' : '0', storageObj.offset),
+        },
+      ];
+    } else if (variableType.label.startsWith('bytes')) {
+      if (!ethers.utils.isHexString(variable, parseInt(variableType.numberOfBytes, 10))) {
+        throw new Error(`invalid bytesN type`);
+      }
+
+      return [
+        {
+          key: slotKey,
+          val: padBytesHexSlotValue(remove0x(variable).padEnd(parseInt(variableType.numberOfBytes, 10) * 2, '0'), storageObj.offset),
+        },
+      ];
+    } else if (variableType.label.startsWith('uint')) {
+      if (remove0x(BigNumber.from(variable).toHexString()).length / 2 > parseInt(variableType.numberOfBytes, 10)) {
+        throw new Error(`provided ${variableType.label} is too big: ${variable}`);
+      }
+
+      return [
+        {
+          key: slotKey,
+          val: padNumHexSlotValue(variable, storageObj.offset),
+        },
+      ];
+    } else if (variableType.label.startsWith('int')) {
```

Comment:
Hmmm is this exactly identical to `uint`? If so, we should just combine into a single block (if label startswith uint or int)

Reply:
Good catch, go for it!